server/test/integration/build_event_protocol/build_event_protocol_test.go

```diff
@@ -28,3 +43,306 @@ func TestBuildWithBESFlags_Success(t *testing.T) {
+// a Bazel event with the same payload type as the given message is successfully
+// forwarded to the build event server. The given message must be assignable to
+// BuildEvent.Payload, otherwise the test immediately fails.
+func AfterForwardBazelEvent(t *testing.T, payloadMsg interface{}) StreamErrorInjector {
+	payloadType := reflect.TypeOf(payloadMsg)
+	payloadSuperType := reflect.TypeOf(&(&bespb.BuildEvent{}).Payload).Elem()
+	if !payloadType.AssignableTo(payloadSuperType) {
+		require.FailNowf(t, "invalid argument", "%s is not assignable to %s", payloadType, payloadSuperType)
+		return nil
+	}
+	return func(event *StreamEvent) error {
+		if !event.WasForwarded || event.ClientRecv == nil || event.ClientRecv.req == nil {
+			return nil
+		}
+		req := event.ClientRecv.req
+		bazelEvent, err := readBazelEvent(req.GetOrderedBuildEvent())
+		if err != nil {
+			return err
+		}
+		if bazelEvent == nil {
+			return nil
+		}
+		if reflect.TypeOf(bazelEvent.GetPayload()).AssignableTo(reflect.TypeOf(payloadType)) {
+			return status.UnavailableError("Proxy: Injected error for test")
+		}
+		return nil
+	}
+}
+
+// BEPProxyServer sits in between Bazel and the build event server, proxying
+// messages directly. It allows injecting errors into the stream to ensure that
+// they are handled properly.
+type BEPProxyServer struct {
+	t             *testing.T
+	backend       pepb.PublishBuildEventClient
+	port          int
+	errorInjector StreamErrorInjector
+}
+
+func StartBEPProxy(t *testing.T, backend pepb.PublishBuildEventClient) *BEPProxyServer {
+	proxy := &BEPProxyServer{
+		t:       t,
+		backend: backend,
+	}
+
+	lis, err := net.Listen("tcp", "0.0.0.0:0")
+	require.NoError(t, err, "failed to start BES proxy")
+	port := lis.Addr().(*net.TCPAddr).Port
+
+	server := grpc.NewServer()
+	pepb.RegisterPublishBuildEventServer(server, proxy)
+	go func() {
+		t.Logf("Proxy: Listening on 0.0.0.0:%d", port)
+		err := server.Serve(lis)
+		require.NoError(t, err)
+	}()
+	t.Cleanup(func() {
+		server.GracefulStop()
+	})
+
+	proxy.port = port
+
+	return proxy
+}
+
+func (p *BEPProxyServer) GRPCAddress() string {
+	return fmt.Sprintf("grpc://localhost:%d", p.port)
+}
+
+// FailOncec registers an error injector that can fire at most once for this
+// proxy server instance. In other words, it is unregistered after returning an
+// error for the first time.
+func (p *BEPProxyServer) FailOnce(f StreamErrorInjector) {
+	p.errorInjector = f
+}
+
+func (p *BEPProxyServer) maybeInjectError(event *StreamEvent) error {
+	if p.errorInjector == nil {
+		return nil
+	}
+	if err := p.errorInjector(event); err != nil {
+		// Clear the error injector after it injects an error, to ensure we inject
+		// the error only once per stream proxy instance.
+		p.errorInjector = nil
+		return err
+	}
+	return nil
+}
+
+func (p *BEPProxyServer) PublishLifecycleEvent(ctx context.Context, req *pepb.PublishLifecycleEventRequest) (*empty.Empty, error) {
+	p.t.Log("Proxy: /PublishLifecycleEvent")
```

Comment:
I would not recommend using the test logger outside of unit tests. 
The test log lines are printed separately from the rest of the logs and do not have timestamps which can make order of events hard to discern from test logs.

Reply:
Ah right, I was running with `-test.v` which shows the test logs in real time, but without `-v` I guess the logs would have to be printed at the end of the test. Fixed